node/src/p2p.rs

```diff
@@ -189,66 +211,91 @@ impl Worker {
                 BehaviourEvent::KeepAlive(_) => {}
                 BehaviourEvent::Gossipsub(ev) => self.on_gossip_sub_event(ev).await?,
             },
+            SwarmEvent::ConnectionEstablished { peer_id, .. } => {
+                self.connected_peers.insert(peer_id);
+            }
+            SwarmEvent::ConnectionClosed { peer_id, .. } => {
+                self.connected_peers.remove(&peer_id);
+            }
             _ => {}
         }
 
         Ok(())
     }
 
-    async fn on_command(&mut self, cmd: P2pCmd) -> Result<()> {
+    async fn on_cmd(&mut self, cmd: P2pCmd) -> Result<()> {
         trace!("{cmd:?}");
 
         match cmd {
             P2pCmd::NetworkInfo { respond_to } => {
-                let _ = respond_to.send(self.swarm.network_info());
+                respond_to.maybe_send(self.swarm.network_info());
+            }
+            P2pCmd::ExchangeHeaderRequest {
+                request,
+                respond_to,
+            } => {
+                self.on_exchange_header_request_cmd(request, respond_to)
+                    .await;
             }
         }
 
         Ok(())
     }
 
-    async fn on_identify_event(&mut self, _ev: &identify::Event) -> Result<()> {
+    async fn on_identify_event(&mut self, _ev: identify::Event) -> Result<()> {
         // TODO
         Ok(())
     }
 
-    async fn on_header_ex_event(&mut self, ev: &exchange::Event) -> Result<()> {
+    async fn on_header_ex_event(&mut self, ev: exchange::Event) -> Result<()> {
         match ev {
-            request_response::Event::Message {
-                peer,
+            exchange::Event::Message {
                 message:
-                    request_response::Message::Response {
+                    exchange::Message::Response {
                         request_id,
                         response,
                     },
+                ..
             } => {
-                println!(
-                    "Response for request: {request_id}, from peer: {peer}, status: {:?}",
-                    response.status_code()
-                );
-                let header = ExtendedHeader::decode(&response.body[..]).unwrap();
-                // TODO: Forward response back with one shot channel
-                println!("Header: {header:?}");
+                let res = match response.status_code() {
+                    StatusCode::Invalid => Err(P2pError::UnsupportedHeaderResponse),
+                    StatusCode::NotFound => Err(P2pError::HeaderNotFound),
+                    StatusCode::Ok => ExtendedHeader::decode(&response.body[..])
+                        .map_err(|_| P2pError::UnsupportedHeaderResponse),
+                };
+
+                trace!("Response: {res:?}");
+
+                let respond_to = self.exchange_reqs.remove(&request_id).unwrap();
+                respond_to.maybe_send(res);
+            }
+
+            exchange::Event::InboundFailure { request_id, .. }
```

Comment:
iirc InboundFailure is for when we fail to receive request from someone, in case of us making requests only outboundfailure should be considered